core/io/primitives.hpp

```diff
@@ -126,6 +208,77 @@ void Gadgetron::Core::IO::read(std::istream &stream, Gadgetron::hoNDArray<T> &ar
     IO::read(stream,array.data(),array.size());
 
 }
+
+template<class T, unsigned int D>
+void Gadgetron::Core::IO::read(std::istream &stream, Gadgetron::hoNDImage<T, D> &array) {
+
+    typedef typename Gadgetron::hoNDImage<T, D>::coord_type coord_type;
+    typedef typename Gadgetron::hoNDImage<T, D>::axis_type axis_type;
+
+    std::vector<size_t> dimensions;
+    std::vector<coord_type> pixelSize, origin, axis_values;
+
+    Gadgetron::Core::IO::read(stream, dimensions);
+    Gadgetron::Core::IO::read(stream, pixelSize);
+    Gadgetron::Core::IO::read(stream, origin);
+    Gadgetron::Core::IO::read(stream, axis_values);
+
+    axis_type axis(D);
+    for (auto d=0; d<D; d++) {
+        for (auto a=0; a<D; a++) {
+            axis[d][a] = axis_values[a + d*D]; 
+        }
+    }
+
+    array = Gadgetron::hoNDImage<T, D>(dimensions, pixelSize, origin, axis);
+    Gadgetron::Core::IO::read(stream, array.data(), array.size());
+}
+
+template<class T, unsigned int D>
+void Gadgetron::Core::IO::read(std::istream &stream, Gadgetron::hoMRImage<T, D> &array) {
+
+    typedef typename Gadgetron::hoMRImage<T, D>::coord_type coord_type;
+    typedef typename Gadgetron::hoMRImage<T, D>::axis_type axis_type;
+
+    std::vector<size_t> dimensions;
+    std::vector<coord_type> pixelSize, origin, axis_values;
+
+    Gadgetron::Core::IO::read(stream, dimensions);
+    Gadgetron::Core::IO::read(stream, pixelSize);
+    Gadgetron::Core::IO::read(stream, origin);
+    Gadgetron::Core::IO::read(stream, axis_values);
+
+    axis_type axis(D);
+    for (auto d=0; d<D; d++) {
+        for (auto a=0; a<D; a++) {
+            axis[d][a] = axis_values[a + d*D]; 
+        }
+    }
+
+    array = Gadgetron::hoMRImage<T, D>(dimensions, pixelSize, origin, axis);
+
+    stream.read(reinterpret_cast<char *>(&array.header_), sizeof(ISMRMRD::ISMRMRD_ImageHeader));
```

Comment:
`ISMRMRD::ISMRMRD_ImageHeader` is the C type; the C++ type is `ISMRMRD::ImageHeader`, I believe. Why this?

Reply:
Fixed